src/Advanced addition, mutiplication et inequality world.tex

```diff
@@ -0,0 +1,101 @@
+\textbf{Apply}\\
+Cette technique vous permet de modifier le but sans rajouter de variables : de fait, elle raisonne comme ceci : vous avez pour but un élément de Q. Or vous disposez d'une fonction f : $P \to Q$. De ce fait, pour disposer d'un élément de Q, il vous suffit de disposer d'un élément de P, car f(p) sera dans Q. \textit{apply f,} fait exactement ça, et donc changera le but de Q en P.\\
+
+
+\paragraph{IV : Function World}
+Ce monde nous introduit un outil fondamental de Lean : les fonctions. \\
+Un élément important à remarquer est qu'en Lean, toutes les fonctions sont curryfiées.\\
+
+Voici un exemple de niveau de ce monde, le niveau 6, qui demande de créer une fonction de fonctions assez fastidieuse, et qui utilise le fait que ces fonctions sont curryfiées. \\
+L'énoncé se formule comme ceci : \\
+$(P Q R : Type) : (P \to (Q \to R)) \to ((P \to Q) \to (P \to R))$\\
+La preuve est de fait assez simple : \\
+\textit{intros f g p,} -- On introduit les différents éléments/fonctions pour créer la fonction demandée \\
+\textit{apply f p,} -- On modifie le but à l'aide de la fonction curryfiée \\
+\textit{exact g p,} -- On trouve le résultat demandé \\
+Ce qui conclut la preuve.\\
+
+\paragraph{V : Proposition World} 
+Dans ce monde on aborde un aspect fondamental de l'assistant de preuves Lean : une preuve est composée d'implications, et c'est ici que les fonctions prennent toute leur importance : pour montrer que A implique B, il suffit de créer une fonction de A vers B, soit un élément de type $A \to B$.\\
+\\
+Pour illustrer ce point, voici un exemple simple, le tout premier niveau de Proposition World. \\ \\
+\textbf{Lemme} : if P is true and $P \to Q$ is true, then Q is true. \\ \\
+Soit en Lean : \textit{(P Q : Prop) (p : P) (h : $P  Q$) : Q} \\
+Donc, en français, on dispose d'une preuve de P, et d'une fonction de P dans Q (i-e d'un élément de type $P \to Q$), trouvons un élément de type Q (montrons que Q est vrai). \\
+Ce qui se résout tout aussi trivialement : \textit{exact h(p),}.\\
+\\
+Un autre niveau intéressant est le niveau 8, qui propose une preuve du lemme suivant : \\
+\\
+\textbf{Lemme} : ($P \to Q) \to (\lnot Q \to \lnot Q)$. \\
+\\
+En Lean, on demande donc de créer une fonction qui prend une preuve que $P \to Q$ et renvoie une preuve de $\lnot Q \to \lnot Q$.
+\\
+Pour cela, la première étape est de disposer d'une preuve de $P \to Q$ :\\
+\textit{intro f,}\\
+Lean nous demande alors de créer un élément de type $\lnot Q \to \lnot Q$, qui serait l'image de la fonction qu'il nous est demandé de créer.\\
+L'astuce est ensuite de revenir à la définition de $\lnot P$ : $\lnot P$ ssi $P \to false$. On retranscrit cette définition : \\
```

Comment:
écris: `$\lnot P \equiv P \to false`, c'est plus joli et clair. N'utilise pas ssi, ça n'a pas de vrai définition.

Plutôt, ¬P est définitionnellement égal à P → false.

Reply:
(mais aussi équivalent à P → false logiquement.)